### PR TITLE
Implement sensor data uploading in LocalTrajectoryUploader.

### PR DESCRIPTION
**Pull request by cschuet**

### CODE REVIEW

File: cartographer_grpc/local_trajectory_uploader.cc [hunk 1]
```diff
@@ -16,26 +16,130 @@
...
+#include "cartographer/common/make_unique.h"
 #include "glog/logging.h"
 
 namespace cartographer_grpc {
+namespace {
+
+const cartographer::common::Duration kPopTimeout =
+    cartographer::common::FromMilliseconds(100);
+const std::string kFixedFramePoseDataMessageName =
+    "cartographer_grpc.proto.AddFixedFramePoseDataRequest";
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer_grpc/local_trajectory_uploader.cc]: As discussed lets go with dynamic_casts then.

File: cartographer_grpc/local_trajectory_uploader.cc [hunk 2]
```diff
@@ -16,26 +16,130 @@
...
+
+void LocalTrajectoryUploader::Shutdown() {
+  CHECK(!shutting_down_);
+  CHECK(upload_thread_);
+  shutting_down_ = true;
+  upload_thread_->join();
+}
+
+void LocalTrajectoryUploader::ProcessSendQueue() {
+  LOG(INFO) << "Starting SLAM thread.";
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer_grpc/local_trajectory_uploader.cc]: done

File: cartographer_grpc/local_trajectory_uploader.cc [hunk 3]
```diff
@@ -16,26 +16,130 @@
...
+  while (!shutting_down_) {
+    auto data_message = send_queue_.PopWithTimeout(kPopTimeout);
+    if (data_message) {
+      const std::string type_name = data_message->GetTypeName();
+      if (type_name == kFixedFramePoseDataMessageName) {
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer_grpc/local_trajectory_uploader.cc]: Done.

File: cartographer_grpc/local_trajectory_uploader.h [hunk 4]
```diff
@@ -31,6 +34,15 @@ namespace cartographer_grpc {
 class LocalTrajectoryUploader {
  public:
   LocalTrajectoryUploader(const std::string& server_address);
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer_grpc/local_trajectory_uploader.h]: Done.

File: cartographer_grpc/local_trajectory_uploader.h [hunk 5]
```diff
@@ -41,12 +53,25 @@ class LocalTrajectoryUploader {
...
   std::shared_ptr<grpc::Channel> client_channel_;
   std::unique_ptr<proto::MapBuilderService::Stub> service_stub_;
   std::map<int, int> local_to_cloud_trajectory_id_map_;
   cartographer::common::BlockingQueue<
       std::unique_ptr<google::protobuf::Message>>
       send_queue_;
+  bool shutting_down_ = false;
+  std::unique_ptr<std::thread> upload_thread_;
+  mapping::ClientWriter<proto::AddImuDataRequest> imu_writer_;
+  mapping::ClientWriter<proto::AddOdometryDataRequest> odometry_writer_;
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer_grpc/local_trajectory_uploader.h]: Done.

File: cartographer_grpc/mapping/client_writer.h [hunk 6]
```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2017 The Cartographer Authors
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer_grpc/mapping/client_writer.h]: Done.

File: cartographer_grpc/mapping/client_writer.h [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+#ifndef CARTOGRAPHER_GRPC_MAPPING_CLIENT_WRITER_H_
+#define CARTOGRAPHER_GRPC_MAPPING_CLIENT_WRITER_H_
+
+#include <memory>
+
+#include "google/protobuf/empty.pb.h"
+#include "grpc++/grpc++.h"
+
+namespace cartographer_grpc {
+namespace mapping {
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer_grpc/mapping/client_writer.h]: Done.

File: cartographer_grpc/mapping/trajectory_builder_stub.h [hunk 8]
```diff
@@ -73,10 +68,10 @@ class TrajectoryBuilderStub
   std::shared_ptr<grpc::Channel> client_channel_;
   const int trajectory_id_;
   std::unique_ptr<proto::MapBuilderService::Stub> stub_;
-  SensorClientWriter<proto::AddRangefinderDataRequest> rangefinder_writer_;
-  SensorClientWriter<proto::AddImuDataRequest> imu_writer_;
-  SensorClientWriter<proto::AddOdometryDataRequest> odometry_writer_;
-  SensorClientWriter<proto::AddFixedFramePoseDataRequest> fixed_frame_writer_;
+  ClientWriter<proto::AddRangefinderDataRequest> rangefinder_writer_;
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer_grpc/mapping/trajectory_builder_stub.h]: Ack.

File: cartographer_grpc/mapping/client_writer.h [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+template <typename RequestType>
+struct ClientWriter {
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer_grpc/mapping/client_writer.h]: See below.

**Comment by cschuet** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user cschuet. Merge queue now has a length of 1.

**Pull request merged**